code/modules/mob/living/status_procs.dm

```diff
@@ -291,7 +291,7 @@
 /////////////////////////////////// SLEEPING ////////////////////////////////////
 
 /mob/living/proc/IsSleeping() //If we're asleep
-	return has_status_effect(STATUS_EFFECT_SLEEPING)
+	return has_status_effect(STATUS_EFFECT_SLEEPING) && !HAS_TRAIT(src, TRAIT_SLEEPIMMUNE)
```

Comment:
Ready for suffering? has_status_effect() does not in fact return a boolean yes or no as the name implies, it returns the trait itself. so &&'ng it with a bool check will cause undesired behavior. Maybe 
```suggestion
  if(!HAS_TRAIT(src, TRAIT_SLEEPIMMUNE))
    return has_status_effect(STATUS_EFFECT_SLEEPING)
```